hal/explicit_pinmap.cpp

```diff
@@ -0,0 +1,105 @@
+
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2019 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "mbed_error.h"
+#include "explicit_pinmap.h"
+
+#if DEVICE_SPI
+MBED_WEAK void spi_init_direct(spi_t *obj, const spi_pinmap_t *pinmap)
```

Comment:
I was thinking about this, but the intention of using direct API is to save memory. Otherwise, the user can use the traditional API. So I assumed that we should inform the user that at the moment this feature is not supported (instead building the program without any information and no memory savings).

Reply:
Yes, the other API can be more efficient, but that's not a reason to not support it when it's only _equally_ efficient. Portable+efficient code (like that mbed_retarget.cpp) isn't going to be happy having to have `#ifdef` around every device usage.